README.md
# Habitat Cookbook

[![Cookbook Version](https://img.shields.io/cookbook/v/habitat.svg?style=flat)](https://supermarket.chef.io/cookbooks/habitat)
[![CI State](https://github.com/chef-cookbooks/habitat/workflows/delivery/badge.svg)](https://github.com/chef-cookbooks/habitat/actions?query=workflow%3Adelivery)
[![License](https://img.shields.io/badge/License-Apache%202.0-green.svg)](https://opensource.org/licenses/Apache-2.0)

## Description

This cookbook provides resources for working with [Habitat](https://habitat.sh). It is intended that these resources will be included in core Chef at some point in the future, so it is important to note:

- APIs are subject to change
- Habitat is a rapidly changing product, and this cookbook may change rapidly as well

## License Note

Habitat requires acceptance of a license before any habitat commands can be run. To accept the Habitat license using this cookbook, the `license` parameter can be set to `accept` for either the `hab_install` or `hab_sup` resources as shown in the below examples:

```ruby
hab_install 'install habitat' do
  license 'accept'
end
```

```ruby
hab_sup 'default' do
  license 'accept'
end
```

PLEASE NOTE: Without performing one of the above license acceptance steps, all other resources in the habitat cookbook will fail with an error prompting that the license must be accepted.

## Requirements

### Platforms

- RHEL 6+
- Ubuntu 16.04+
- Debian 9+
- Windows 2016+

### Habitat

- Habitat version: 1.5.71

This cookbook is developed lockstep with the latest release of Habitat to ensure compatibility, going forward from 0.33.0 of the cookbook and 0.33.2 of Habitat itself. When new versions of Habitat are released, the version should be updated in these files:

- `README.md`: note required version in this file
- `resources/install.rb`: set the default to the new version
- `test/integration/install/default_spec.rb`: to match the version from the resource

Additionally, new versions must be tested that all behavior in the cookbook still works, otherwise the cookbook must be updated to match the behavior in the new version of Habitat.

Users who wish to install a specific version of Habitat should use an older (0.28 or earlier) release of this cookbook, but note that is unsupported and they are advised to upgrade ASAP.

### Chef

- Chef 12.20.3+

### Cookbooks

- None

## Resources

### hab_install

Installs Habitat on the system using the [install script](https://raw.githubusercontent.com/habitat-sh/habitat/master/components/hab/install.sh).

#### Actions

- `install`: Installs Habitat. Does nothing if the `hab` binary is found in the default location for the system (`/bin/hab` on Linux, `/usr/local/bin/hab` on macOS, `C:/habitat/hab.exe` on Windows)
- `upgrade`: Installs the latest version of Habitat, does not check if the binary exists

#### Properties

- `install_url`: URL to the install script, default is from the [habitat repo](https://raw.githubusercontent.com/habitat-sh/habitat/master/components/hab/install.sh)
- `bldr_url`: Optional URL to an alternate Builder (defaults to the public Builder)
- `create_user`: Creates the `hab` system user (defaults to `true`)
- `tmp_dir`: Sets TMPDIR environment variable for location to place temp files.  (required if `/tmp` and `/var/tmp` are mounted `noexec`)
- `license`: Specifies acceptance of habitat license when set to `accept` (defaults to empty string).

#### Examples

Nameless installation

```ruby
hab_install
```

Instalaltion specifying a bldr URL

```ruby
hab_install 'install habitat' do
  bldr_url 'http://localhost'
end
```

### hab_package

Install the specified Habitat package from builder. Requires that Habitat is installed

#### actions

- `install`: installs the specified package
- `upgrade`: installs the specified package. If a newer version is available in the configured channel, upgrades to that version

#### Properties

- `package_name`: A Habitat package name, must include the origin and package name separated by `/`, for example, `core/redis`
- `version`: A Habitat version which contains the version and optionally a release separated by `/`, for example, `3.2.3` or `3.2.3/20160920131015`
- `bldr_url`: The habitat builder url where packages will be downloaded from (defaults to public habitat builder)
- `channel`: The release channel to install from (defaults to `stable`)
- `auth_token`: Auth token for installing a package from a private organization on builder
- `binlink`: If habitat should attempt to binlink the package.  Acceptable values: `true`, `false`, `:force`.  Will faill on binlinking if set to `true` and binary or binlink exists.
- `options`: Pass any additional parameters to the habitat install command.

While it is valid to pass the version and release with a Habitat package as a fully qualified package identifier when using the `hab` CLI, they must be specified using the `version` property when using this resource. See the examples below.

#### Examples

```ruby
hab_package 'core/redis'

hab_package 'core/redis' do
  version '3.2.3'
  channel 'unstable'
end

hab_package 'core/redis' do
  version '3.2.3/20160920131015'
end

hab_package 'core/nginx' do
  binlink :force
end

hab_package 'core/nginx' do
  options '--binlink'
end
```

### hab_service

Manages a Habitat application service using `hab sup`/`hab service`. This requires that `core/hab-sup` be running as a service. See the `hab_sup` resource documentation below for more information about how to set that up with this cookbook.

_Note:_ Applications may run as a specific user. Often with Habitat, the default is `hab`, or `root`. If the application requires another user, then it should be created with Chef's `user` resource.

#### Actions

- `:load`: (default action) runs `hab service load` to load and start the specified application service
- `:unload`: runs `hab service unload` to unload and stop the specified application service
- `:reload`: runs the `:unload` and then `:load` actions
- `:start`: runs `hab service start` to start the specified application service
- `:stop`: runs `hab service stop` to stop the specified application service
- `:restart`: runs the `:stop` and then `:start` actions

#### Properties

The remote_sup property is valid for all actions.

- `remote_sup`: Address to a remote Supervisor's Control Gateway [default: 127.0.0.1:9632]
- `remote_sup_http`: Address for remote supervisor http port. Used to pull existing configuration data. If this is invalid, config will be applied on every Chef run.
- `gateway_auth_token`: Auth token for accessing the remote supervisor's http port.

The follow properties are valid for the `load` action.

- `service_name`: name property, the name of the service, must be in the form of `origin/name`
- `loaded`: state property indicating whether the service is loaded in the supervisor
- `running`: state property indicating whether the service is running in the supervisor
- `strategy`: Passes `--strategy` with the specified update strategy to the hab command
- `update_condition`: Passes `--update-condition` dictating when this service should updated. Defaults to `latest`. Options are `latest` or `track-channel` ***Note: This requires a minimum habitat version of 1.5.71***
  - `latest`: Runs the latest package that can be found in the configured channel and local packages.
  - `track-channel`: Always run what is at the head of a given channel. This enables service rollback where demoting a package from a channel will cause the package to rollback to an older version of the package. A ramification of enabling this condition is packages newer than the package at the head of the channel will be automatically uninstalled during a service rollback.
- `topology`: Passes `--topology` with the specified service topology to the hab command
- `bldr_url`: Passes `--url` with the specified Builder URL to the hab command.
- `channel`: Passes `--channel` with the specified channel to the hab command
- `bind`: Passes `--bind` with the specified services to bind to the hab command. If an array of multiple service binds are specified then a `--bind` flag is added for each.
- `binding_mode`: Passes `--binding-mode` with the specified binding mode. Defaults to `:strict`. Options are `:strict` or `:relaxed`
- `service_group`: Passes `--group` with the specified service group to the hab command

#### Examples

```ruby
# install and load nginx
hab_package 'core/nginx'
hab_service 'core/nginx'

hab_service 'core/nginx unload' do
  service_name 'core/nginx'
  action :unload
end

# pass the strategy and topology options to hab service commands (load by default)
hab_service 'core/redis' do
  strategy 'rolling'
  topology 'standalone'
end

# Using update_condition
hab_service 'core/redis' do
  strategy 'rolling'
  update_condition 'track-channel'
  topology 'standalone'
end
```

If the service has it's own user specified that is not the `hab` user, don't create the `hab` user on install, and instead create the application user with Chef's `user` resource

```ruby
hab_install 'install habitat' do
  create_user false
end

user 'acme-apps' do
  system true
end

hab_service 'acme/apps'
```

### hab_sup

Runs a Habitat Supervisor for one or more Habitat Services. It is used in conjunction with `hab_service` which will manage the services loaded and started within the supervisor.

The `run` action handles installing Habitat using the `hab_install` resource, ensures that the appropriate versions of the `core/hab-sup` and `core/hab-launcher` packages are installed using `hab_package`, and then drops off the appropriate init system definitions and manages the service.

All `event_stream_*` properties are optional, and allow the Habitat Supervisor to display details about it's status and running services via the [Chef Automate Applications Dashboard](https://automate.chef.io/docs/applications-dashboard/).
***Note: Automate has TLS on by default. You will need to follow these instructions to make sure habitat has the proper certificates for `event_stram_*` [Share the TLS Certificate with Chef Habitat](https://automate.chef.io/docs/applications-setup/#share-the-tls-certificate-with-chef-habitat)***

#### Actions

- `run`: starts the `hab-sup` service

#### Properties

- `bldr_url`: The Builder URL for the `hab_package` resource, if needed
- `permanent_peer`: Only valid for `:run` action, passes `--permanent-peer` to the hab command
- `listen_ctl`: Only valid for `:run` action, passes `--listen-ctl` with the specified address and port, e.g., `0.0.0.0:9632`, to the hab command
- `listen_gossip`: Only valid for `:run` action, passes `--listen-gossip` with the specified address and port, e.g., `0.0.0.0:9638`, to the hab command
- `listen_http`: Only valid for `:run` action, passes `--listen-http` with the specified address and port, e.g., `0.0.0.0:9631`, to the hab command
- `org`: Only valid for `:run` action, passes `--org` with the specified org name to the hab command
- `peer`: Only valid for `:run` action, passes `--peer` with the specified initial peer to the hab command
- `ring`: Only valid for `:run` action, passes `--ring` with the specified ring key name to the hab command
- `auto_update`: Passes `--auto-update`. This will set the Habitat supervisor to automatically update itself any time a stable version has been released
- `update_condition`: Passes `--update-condition` dictating when this service should updated. Defaults to `latest`. Options are `latest` or `track-channel`  ***Note: This requires a minimum habitat version of 1.5.71***
  - `latest`: Runs the latest package that can be found in the configured channel and local packages.
  - `track-channel`: Always run what is at the head of a given channel. This enables service rollback where demoting a package from a channel will cause the package to rollback to an older version of the package. A ramification of enabling this condition is packages newer than the package at the head of the channel will be automatically uninstalled during a service rollback.
- `hab_channel`: The channel to install Habitat from. Defaults to stable
- `auth_token`: Auth token for accessing a private organization on bldr. This value is templated into the appropriate service file.
- `gateway_auth_token`: Auth token for accessing the supervisor's HTTP gateway. This value is templated into the appropriate service file.
- `license`: Specifies acceptance of habitat license when set to `accept` (defaults to empty string).
- `health_check_interval`: The interval (seconds) on which to run health checks (defaults to 30).
- `event_stream_application`: The name of your application that will be displayed in the Chef Automate Applications Dashboard
- `event_stream_environment`: The application environment for the supervisor, this is for grouping in the Applications Dashboard
- `event_stream_site`: Application Dashboard label for the "site" of the application - can be filtered in the dashboard
- `event_stream_url`: `AUTOMATE_HOSTNAME:4222` - the Chef Automate URL with port 4222 specified (can be changed if needed)
- `event_stream_token`: Chef Automate token for sending application event stream data
- `event_stream_certificate`: With `Intermediary Certificates` or, Automate 2 being set to use TLS with a valid cert, you will need to provide `Habitat` with your certificate for communication with Automate to work. [Follow these steps!](https://automate.chef.io/docs/applications-setup/#share-the-tls-certificate-with-chef-habitat)

#### Examples

```ruby
# set up with just the defaults
hab_sup 'default'

hab_sup 'test-options' do
  listen_http '0.0.0.0:9999'
  listen_gossip '0.0.0.0:9998'
end

# Use with an on-prem Builder
# Access to public builder may not be available
hab_sup 'default' do
  bldr_url 'https://bldr.private.net'
end

# Using update_condition
hab_sup 'default' do
  bldr_url 'https://bldr.private.net'
  hab_channel 'dev'
  update_condition 'track-channel'
end

# Provide event_stream_* information
hab_sup 'default' do
  license 'accept'
  event_stream_application 'myapp'
  event_stream_environment 'production'
  event_stream_site 'MySite'
  event_stream_url 'automate.private.net:4222'
  event_stream_token 'myawesomea2clitoken='
end

```

### hab_config

Applies a given configuration to a habitat service using `hab config apply`.

#### Actions

- `apply`: (default action) apply the given configuration

#### Properties

- `service_group`: The service group to apply the configuration to, for example, `nginx.default`
- `config`: The configuration to apply as a ruby hash, for example, `{ worker_count: 2, http: { keepalive_timeout: 120 } }`
- `remote_sup`: Address to a remote Supervisor's Control Gateway [default: 127.0.0.1:9632]
- `remote_sup_http`: Address for remote supervisor http port. Used to pull existing configuration data. If this is invalid, config will be applied on every Chef run.
- `gateway_auth_token`: Auth token for accessing the remote supervisor's http port.
- `user`: Name of user key to use for encryption. Passes `--user` to `hab config apply`

#### Notes

The version number of the configuration is automatically generated and will be the current timestamp in seconds since 1970-01-01 00:00:00 UTC.

#### Examples

```ruby
hab_config 'nginx.default' do
  config({
    worker_count: 2,
    http: {
      keepalive_timeout: 120
    }
  })
end
```

### hab_user_toml

Templates a user.toml for the specified service. This is written to `/hab/user/<service_name>/config/user.toml`. User.toml can be used to set configuration overriding the default.toml for a given package as an alternative to applying service group level configuration.

#### Actions

- `create`: (default action) Create the user.toml from the specified config.
- `delete`: Delete the user.toml

#### Properties

- `service_name`: The service group to apply the configuration to, for example, `nginx.default`
- `config`: Only valid for `:create` action. The configuration to apply as a ruby hash, for example, `{ worker_count: 2, http: { keepalive_timeout: 120 } }`

#### Examples

```ruby
hab_user_toml 'nginx' do
  config({
    worker_count: 2,
    http: {
      keepalive_timeout: 120
    }
  })
end
```

## Maintainers

This cookbook is maintained by the following maintainers:

- Jon Cowie [jcowie@chef.io](mailto:jcowie@chef.io)
- Jeff Brimager [jbrimager@chef.io](mailto:jbrimager@chef.io)

The goal of the Community Cookbook Engineering team is to improve cookbook quality and to aid the community in contributing to cookbooks. To learn more about our team, process, and design goals see our [team documentation](https://github.com/chef-cookbooks/community_cookbook_documentation/blob/master/COOKBOOK_TEAM.MD). To learn more about contributing to cookbooks like this see our [contributing documentation](https://github.com/chef-cookbooks/community_cookbook_documentation/blob/master/CONTRIBUTING.MD), or if you have general questions about this cookbook come chat with us in #cookbok-engineering on the [Chef Community Slack](http://community-slack.chef.io/)

## License

**Copyright:** 2016-2018, Chef Software, Inc.

```
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
# Habitat Cookbook CHANGELOG

This file is used to list changes made in each version of the habitat cookbook.
 <!-- latest_release unreleased -->

- resolved cookstyle error: libraries/provider_hab_package.rb:174:13 convention: `Style/RedundantReturn`
- resolved cookstyle error: libraries/provider_hab_package.rb:176:13 convention: `Style/RedundantReturn`

<!-- latest_release -->
## 1.7.0 (2020-04-08)

#### Merged Pull Requests

- Automated PR: Standardising Files [#226](https://github.com/chef-cookbooks/habitat/pull/226) ([xorimabot](https://github.com/xorimabot))
- Automated PR: Cookstyle Changes [#225](https://github.com/chef-cookbooks/habitat/pull/225) ([xorimabot](https://github.com/xorimabot))
- adds windows Github Actions testing for issue #193 [#224](https://github.com/chef-cookbooks/habitat/pull/224) ([collinmcneese](https://github.com/collinmcneese))
- changing the windows service cofig to better reflect newer functional… [#223](https://github.com/chef-cookbooks/habitat/pull/223) ([sam1el](https://github.com/sam1el))
- removing update-confition defaults and tests [#221](https://github.com/chef-cookbooks/habitat/pull/221) ([sam1el](https://github.com/sam1el))
- Added optional settings for Windows service to recongize local or provided envoringment variables including
  - HAB_AUTH_TOKEN
  - HAB_GATEWAY_AUTH_TOKEN
  - HAB_BLDR_URL

## 1.6.2 (2020-04-07)

#### Merged Pull Requests

- changing the windows service cofig to better reflect newer functional… [#223](https://github.com/chef-cookbooks/habitat/pull/223) ([sam1el](https://github.com/sam1el))
- removing update-confition defaults and tests [#221](https://github.com/chef-cookbooks/habitat/pull/221) ([sam1el](https://github.com/sam1el))


## 1.6.0 (2020-04-02)

#### Merged Pull Requests	- complete overhaul of the windows-service config to refplect current best practice

- adding --update-condition funtionality that was added in habitat 1.5.71 [#217](https://github.com/chef-cookbooks/habitat/pull/217) ([sam1el](https://github.com/sam1el))
- added --update-condition funtionality that was added in habitat 1.5.71. See README.MD for info on usage
- complete overhaul of the windows-service config to refplect current best practice


## 1.5.10 (2020-04-01)

#### Merged Pull Requests

- Versioning [#215](https://github.com/chef-cookbooks/habitat/pull/215) ([sam1el](https://github.com/sam1el))

## 1.5.9 (2020-03-30)

#### Merged Pull Requests

- updates to be backwards compatible with 12.20.3+ [#214](https://github.com/chef-cookbooks/habitat/pull/214) ([collinmcneese](https://github.com/collinmcneese))

## 1.5.8 (2020-03-09)

#### Merged Pull Requests

- fixing idempotence on hab_sup_systemd. [#211](https://github.com/chef-cookbooks/habitat/pull/211) ([sam1el](https://github.com/sam1el))

## 1.5.7 (2020-03-06)

- Hab version bumped to 1.5.50

#### Merged Pull Requests

- Hab version [#210](https://github.com/chef-cookbooks/habitat/pull/210) ([sam1el](https://github.com/sam1el))

## 1.5.6 (2020-03-02)

- Cookstyle fixes - [@Xorima](https://github.com/Xorima)
- opensuse requires gzip to be installer for hab install script - [@Xorima](https://github.com/Xorima)
- fix readme - [@Xorima](https://github.com/Xorima)
- updated hab sup resource to include new Application Dashboard options (#206) - [@danielcbright](https://github.com/danielcbright)

#### Merged Pull Requests

- Delete Merged Branches [#209](https://github.com/chef-cookbooks/habitat/pull/209) ([Xorima](https://github.com/Xorima))
- Use Github actions badge in the readme [#208](https://github.com/chef-cookbooks/habitat/pull/208) ([Xorima](https://github.com/Xorima))
- Migrate testing to Github Actions and add support for installing on SUSE [#207](https://github.com/chef-cookbooks/habitat/pull/207) ([Xorima](https://github.com/Xorima))

## 1.5.5 (2020-02-26)

#### Merged Pull Requests

- Extending Windows chef-client compatibility  [#205](https://github.com/chef-cookbooks/habitat/pull/205) ([sam1el](https://github.com/sam1el))

## 1.5.4 (2020-02-25)

#### Merged Pull Requests

- adding delay loop to the load function in the service resouce [#204](https://github.com/chef-cookbooks/habitat/pull/204) ([sam1el](https://github.com/sam1el))

## 1.5.3 (2020-02-24)

- Fix linux installer issues

#### Merged Pull Requests

- [BUG FIX] Linux installers [#203](https://github.com/chef-cookbooks/habitat/pull/203) ([sam1el](https://github.com/sam1el))

## 1.5.2 (2020-02-20)

- Fix windows habitat install

#### Merged Pull Requests

- [BUG FIX] Windows install to correct issue #199 [#202](https://github.com/chef-cookbooks/habitat/pull/202) ([sam1el](https://github.com/sam1el))
- Additional cookstyle fixes [#189](https://github.com/chef-cookbooks/habitat/pull/189) ([tas50](https://github.com/tas50))

## 1.5.1 (2020-01-31)

- Fix windows package download URL to use packages.chef.io

#### Merged Pull Requests

- Fix windows package download URL [#198](https://github.com/chef-cookbooks/habitat/pull/198) ([jonlives](https://github.com/jonlives))

## [1.5.0](https://github.com/chef-cookbooks/habitat/tree/1.5.0) (2020-01-29)

- Update cookbook to Habitat 1.5.0
- Fix habitat installs on Windows when using installation archives

#### Merged Pull Requests

- WIP: Update to hab 1.5 and update launchers and windows-service pins [#195](https://github.com/chef-cookbooks/habitat/pull/195) ([jonlives](https://github.com/jonlives))
- Fix Habitat install on windows [#194](https://github.com/chef-cookbooks/habitat/pull/194) ([emachnic](https://github.com/emachnic))

## 0.88.2 (2020-01-24)

- Add support for supervisor HTTP gateway auth token
- Service Properties, Reload/Restart Fixes, Idempotence Improvements

#### Merged Pull Requests

- Service Properties, Reload/Restart Fixes, Idempotence Improvements [#187](https://github.com/chef-cookbooks/habitat/pull/187) ([sirajrauff](https://github.com/sirajrauff))

test/solo.rb
cookbook_path ['test/fixtures/cookbooks', '../']


resources/service.rb
# Copyright:: 2017-2018 Chef Software, Inc.
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

resource_name :hab_service

property :service_name, String, name_property: true
property :loaded, [true, false], default: false
property :running, [true, false], default: false

# hab svc options which get included based on the action of the resource
property :strategy, [Symbol, String], equal_to: [:none, 'none', :atonce, 'at-once', :rolling, 'rolling'], default: :none, coerce: proc { |s| s.is_a?(String) ? s.to_sym : s }
property :topology, [Symbol, String], equal_to: [:standalone, 'standalone', :leader, 'leader'], default: :standalone, coerce: proc { |s| s.is_a?(String) ? s.to_sym : s }
property :bldr_url, String, default: 'https://bldr.habitat.sh'
property :channel, [Symbol, String], default: :stable, coerce: proc { |s| s.is_a?(String) ? s.to_sym : s }
property :bind, [String, Array], coerce: proc { |b| b.is_a?(String) ? [b] : b }, default: []
property :binding_mode, [Symbol, String], equal_to: [:strict, 'strict', :relaxed, 'relaxed'], default: :strict, coerce: proc { |s| s.is_a?(String) ? s.to_sym : s }
property :service_group, String, default: 'default'
property :shutdown_timeout, Integer, default: 8
property :health_check_interval, Integer, default: 30
property :remote_sup, String, default: '127.0.0.1:9632', desired_state: false
# Http port needed for querying/comparing current config value
property :remote_sup_http, String, default: '127.0.0.1:9631', desired_state: false
property :gateway_auth_token, String, desired_state: false
property :update_condition, String

load_current_value do
  service_details = get_service_details(service_name)

  running service_up?(service_details)
  loaded service_loaded?(service_details)

  if loaded
    service_name get_spec_identifier(service_details)
    strategy get_update_strategy(service_details)
    update_condition get_update_condition(service_details)
    topology get_topology(service_details)
    bldr_url get_builder_url(service_details)
    channel get_channel(service_details)
    bind get_binds(service_details)
    binding_mode get_binding_mode(service_details)
    service_group get_service_group(service_details)
    shutdown_timeout get_shutdown_timeout(service_details)
    health_check_interval get_health_check_interval(service_details)
  end

  Chef::Log.debug("service #{service_name} service name: #{service_name}")
  Chef::Log.debug("service #{service_name} running state: #{running}")
  Chef::Log.debug("service #{service_name} loaded state: #{loaded}")
  Chef::Log.debug("service #{service_name} strategy: #{strategy}")
  Chef::Log.debug("service #{service_name} update condition: #{update_condition}")
  Chef::Log.debug("service #{service_name} topology: #{topology}")
  Chef::Log.debug("service #{service_name} builder url: #{bldr_url}")
  Chef::Log.debug("service #{service_name} channel: #{channel}")
  Chef::Log.debug("service #{service_name} binds: #{bind}")
  Chef::Log.debug("service #{service_name} binding mode: #{binding_mode}")
  Chef::Log.debug("service #{service_name} service group: #{service_group}")
  Chef::Log.debug("service #{service_name} shutdown timeout: #{shutdown_timeout}")
  Chef::Log.debug("service #{service_name} health check interval: #{health_check_interval}")
end

# This method is defined here otherwise it isn't usable in the
# `load_current_value` method.
#
# It performs a check with TCPSocket to ensure that the HTTP API is
# available first. If it cannot connect, it assumes that the service
# is not running. It then attempts to reach the `/services` path of
# the API to get a list of services. If this fails for some reason,
# then it assumes the service is not running.
#
# Finally, it walks the services returned by the API to look for the
# service we're configuring. If it is "Up", then we know the service
# is running and fully operational according to Habitat. This is
# wrapped in a begin/rescue block because if the service isn't
# present and `sup_for_service_name` will be nil and we will get a
# NoMethodError.
#
def get_service_details(svc_name)
  http_uri = "http://#{remote_sup_http}"

  begin
    TCPSocket.new(URI(http_uri).host, URI(http_uri).port).close
  rescue Errno::ECONNREFUSED, Errno::EHOSTUNREACH
    Chef::Log.debug("Could not connect to #{http_uri} to retrieve status for #{service_name}")
    return false
  end

  begin
    headers = {}
    headers['Authorization'] = "Bearer #{gateway_auth_token}" if property_is_set?(:gateway_auth_token)
    svcs = Chef::HTTP::SimpleJSON.new(http_uri).get('/services', headers)
  rescue
    Chef::Log.debug("Could not connect to #{http_uri}/services to retrieve status for #{service_name}")
    return false
  end

  origin, name, _version, _release = svc_name.split('/')
  svcs.find do |s|
    s['pkg']['origin'] == origin && s['pkg']['name'] == name
  end
end

def service_up?(service_details)
  service_details['process']['state'] == 'up'
rescue
  Chef::Log.debug("#{service_name} not found on the Habitat supervisor")
  false
end

def service_loaded?(service_details)
  if service_details
    true
  else
    false
  end
end

def get_spec_identifier(service_details)
  service_details['spec_ident']['spec_identifier']
rescue
  Chef::Log.debug("#{service_name} not found on the Habitat supervisor")
  nil
end

def get_update_strategy(service_details)
  service_details['update_strategy'].to_sym
rescue
  Chef::Log.debug("Update Strategy for #{service_name} not found on Supervisor API")
  'none'
end

def get_update_condition(service_details)
  service_details['update_condition']['secs']
rescue
  Chef::Log.debug("Update condition #{service_name} not found on Supervisor API")
  'latest'
end

def get_topology(service_details)
  service_details['topology'].to_sym
rescue
  Chef::Log.debug("Topology for #{service_name} not found on Supervisor API")
  'standalone'
end

def get_builder_url(service_details)
  service_details['bldr_url']
rescue
  Chef::Log.debug("Builder URL for #{service_name} not found on Supervisor API")
  'https://bldr.habitat.sh'
end

def get_channel(service_details)
  service_details['channel'].to_sym
rescue
  Chef::Log.debug("Channel for #{service_name} not found on Supervisor API")
  'stable'
end

def get_binds(service_details)
  service_details['binds']
rescue
  Chef::Log.debug("Update Strategy for #{service_name} not found on Supervisor API")
  []
end

def get_binding_mode(service_details)
  service_details['binding_mode'].to_sym
rescue
  Chef::Log.debug("Binding mode for #{service_name} not found on Supervisor API")
  'strict'
end

def get_service_group(service_details)
  service_details['service_group'].split('.').last
rescue
  Chef::Log.debug("Service Group for #{service_name} not found on Supervisor API")
  'default'
end

def get_shutdown_timeout(service_details)
  service_details['pkg']['shutdown_timeout']
rescue
  Chef::Log.debug("Shutdown Timeout for #{service_name} not found on Supervisor API")
  8
end

def get_health_check_interval(service_details)
  service_details['health_check_interval']['secs']
rescue
  Chef::Log.debug("Health Check Interval for #{service_name} not found on Supervisor API")
  30
end

action :load do
  modified = false
  converge_if_changed :service_name do
    modified = true
  end
  converge_if_changed :strategy do
    modified = true
  end
  converge_if_changed :update_condition do
    modified = true
  end
  converge_if_changed :topology do
    modified = true
  end
  converge_if_changed :bldr_url do
    modified = true
  end
  converge_if_changed :channel do
    modified = true
  end
  converge_if_changed :bind do
    modified = true
  end
  converge_if_changed :binding_mode do
    modified = true
  end
  converge_if_changed :service_group do
    modified = true
  end
  converge_if_changed :shutdown_timeout do
    modified = true
  end
  converge_if_changed :health_check_interval do
    modified = true
  end

  options = svc_options
  if current_resource.loaded && modified
    Chef::Log.debug("Reloading #{current_resource.service_name} using --force due to parameter change")
    options << '--force'
  end

  unless current_resource.loaded && !modified
    execute 'test' do
      command "hab svc load #{new_resource.service_name} #{options.join(' ')}"
      retry_delay 10
      retries 5
    end
  end
end

action :unload do
  if current_resource.loaded
    execute "hab svc unload #{new_resource.service_name} #{svc_options.join(' ')}"
    wait_for_service_unloaded
  end
end

action :start do
  unless current_resource.loaded
    Chef::Log.fatal("No service named #{new_resource.service_name} is loaded on the Habitat supervisor")
    raise "No service named #{new_resource.service_name} is loaded on the Habitat supervisor"
  end

  execute "hab svc start #{new_resource.service_name} #{svc_options.join(' ')}" unless current_resource.running
end

action :stop do
  unless current_resource.loaded
    Chef::Log.fatal("No service named #{new_resource.service_name} is loaded on the Habitat supervisor")
    raise "No service named #{new_resource.service_name} is loaded on the Habitat supervisor"
  end

  if current_resource.running
    execute "hab svc stop #{new_resource.service_name} #{svc_options.join(' ')}"
    wait_for_service_stopped
  end
end

action :restart do
  action_stop
  action_start
end

action :reload do
  action_unload
  action_load
end

action_class do
  def svc_options
    opts = []

    # certain options are only valid for specific `hab svc` subcommands.
    case action
    when :load
      opts.push(*new_resource.bind.map { |b| "--bind #{b}" }) if new_resource.bind
      opts << "--binding-mode #{new_resource.binding_mode}"
      opts << "--url #{new_resource.bldr_url}" if new_resource.bldr_url
      opts << "--channel #{new_resource.channel}" if new_resource.channel
      opts << "--group #{new_resource.service_group}" if new_resource.service_group
      opts << "--strategy #{new_resource.strategy}" if new_resource.strategy
      opts << "--update-condition #{new_resource.update_condition}" if new_resource.update_condition
      opts << "--topology #{new_resource.topology}" if new_resource.topology
      opts << "--health-check-interval #{new_resource.health_check_interval}" if new_resource.health_check_interval
      opts << "--shutdown-timeout #{new_resource.shutdown_timeout}" if new_resource.shutdown_timeout
    when :unload, :stop
      opts << "--shutdown-timeout #{new_resource.shutdown_timeout}" if new_resource.shutdown_timeout
    end

    opts << "--remote-sup #{new_resource.remote_sup}" if new_resource.remote_sup

    opts.map(&:split).flatten.compact
  end

  def wait_for_service_unloaded
    ruby_block 'wait-for-service-unloaded' do
      block do
        raise "#{new_resource.service_name} still loaded" if service_loaded?(get_service_details(new_resource.service_name))
      end
      retries get_shutdown_timeout(new_resource.service_name) + 1
      retry_delay 1
    end

    ruby_block 'update current_resource' do
      block do
        current_resource.loaded = service_loaded?(get_service_details(new_resource.service_name))
      end
      action :nothing
      subscribes :run, 'ruby_block[wait-for-service-unloaded]', :immediately
    end
  end

  def wait_for_service_stopped
    ruby_block 'wait-for-service-stopped' do # ~FC009
      block do
        raise "#{new_resource.service_name} still running" if service_up?(get_service_details(new_resource.service_name))
      end
      retries get_shutdown_timeout(new_resource.service_name) + 1
      retry_delay 1

      ruby_block 'update current_resource' do
        block do
          current_resource.running = service_up?(get_service_details(new_resource.service_name))
        end
        action :nothing
        subscribes :run, 'ruby_block[wait-for-service-stopped]', :immediately
      end
    end
  end
end
